astropy/io/fits/hdu/base.py

```diff
@@ -184,8 +184,12 @@ def __init_subclass__(cls, **kwargs):
             # Don't do anything if the class has already explicitly
             # set the deleter for its data property
             def data(self):
-                # The deleter
-                if self._file is not None and self._data_loaded:
+                # The deleter. sys.getrefcount is CPython specific and not on PyPy.
+                if (
+                    self._file is not None
+                    and self._data_loaded
+                    and hasattr(sys, "getrefcount")
+                ):
                     data_refcount = sys.getrefcount(self.data)
                     # Manually delete *now* so that FITS_rec.__del__
                     # cleanup can happen if applicable
```

Comment:
The line below (`del self.__dict__["data"]`) should be called if `getrefcount` isn't available. So maybe something like this instead:
```
if hasattr(sys, "getrefcount"):
    data_refcount = sys.getrefcount(self.data)

del self.__dict__["data"]

if hasattr(sys, "getrefcount") and data_refcount == 2:
    self._file._maybe_close_mmap()

Reply:
@saimn , I think I applied your requested changes. Please re-review. Thanks!